docs/proposals/volumes.md

```diff
@@ -0,0 +1,373 @@
+should never perform ownership or label management for.
+
+#### Volumes derived from empty dir
+
+Empty dir and volumes derived from it are the opposite of hostPath.  Since they are created by the
+system, Kubernetes must always ensure that the ownership and SELinux context (when relevant) are
+set correctly for the volume to be usable.
+
+#### Distributed file systems
+
+Whether the ownership or SELinux label of a volume for a distributed file system should be managed
+by Kubernetes depends on the cluster operator's policy and the volume type.  Additionally, not all
+distributed file systems support `chown`, `chmod`, and `chcon` operations uniformly.  Our API should
+make it simple to express whether a particular volume should have these concerns managed by
+Kubernetes.  For simplicity's sake, we will assume that if a volume requires ownership management,
+that it also requires relabeling and vice-versa.
+
+#### API requirements
+
+From the above, we know that ownership management must be applied:
+
+1.  To some volume types always
+2.  To some volume types never
+3.  To some volume types *sometimes*
+
+For volume types where the need for ownership/label management is not at the discretion of the user,
+there should be no need (or way) to specify whether ownership management is required in the API.
+Conversely there must be a way to specify this for volume types where it depends on the user's
+needs.
+
+### Determining correct ownership
+
+There are two components of ownership management:
+
+1.  When does Kubernetes need to manage the ownership of a volume?
+2.  How does Kubernetes determine the ownership of a volume?
+
+#### When to determine ownership
+
+Whether Kubernetes should manage the ownership of a volume for a distributed filesystem depends upon
+both the file system type and the cluster operator's policy.  For example:
+
+1.  Some organizations will manage ownership of volumes externally to the cluster
+2.  It is not possible to securely `chown` or `chmod` paths within some distributed filesystems
+
+#### How to determine ownership
+
+The Kubelet must analyze the pod spec to determine which UIDs need to use which volumes.  If a
+container's security context's `RunAsUser` field is not set, the Kubelet must inspect the image via
+the container runtime to determine which UID the image will run as.  Once the list of UIDs that need
+to use a volume is known, the kubelet can determine which ownership and permissions should be used
+to make the volume functional.
+
+If a volume is used only by a single UID within the pod, the ownership can be set to that UID.
+Otherwise, the volume should be owned by a group and the containers run in that group.
+
+If a non-numeric user is specified by an image, the behavior of container runtimes is to look up the
+UID from the container's `/etc/passwd` file.  It is not feasible for the kubelet to make this
+determination; we may not be able to correctly support non-numeric users in image metadata.  If a
```

Comment:
I think there will be more of these challenges when trying to determine the container's UID.. GIDs (generated or specified by the user) are the way to go IMO